File: Phase05/SearchEngineLibrary/SearchEngineLibrary/SearchEngine.cs
using System.Collections.Generic;
using System.IO;
using System.Linq;

namespace SearchEngineLibrary
{
    public class SearchEngine : ISearchEngine
    {
        private static IInvertedIndex _invertedIndex;
        
        public SearchEngine(IInvertedIndex invertedIndex)
        {
            _invertedIndex = invertedIndex;
        }

        public HashSet<string> GetResult(string query)
        {
            var result = new HashSet<string>();
            var modifiedQuery = GetModifiedQuery(query);

            var andList = new List<string>();
            var orList = new List<string>();
            var excludeList = new List<string>();
            
            FillListsByQuery(modifiedQuery,andList,orList,excludeList);
            
            AddAndWordsToResult(result,andList);
            AddOrWordsToResult(result,orList);
            RemoveExcludeWordsFromResult(result,excludeList);

            return result;
        }
        
        private static string GetModifiedQuery(string query)
        {
            return query.ToLower();
        }
        
        private static void AddAndWordsToResult(HashSet<string> result, List<string> andList) 
        {
            foreach (string and in andList){
                if (!result.Any()){
                    result.UnionWith(_invertedIndex.GetResultListByWord(and));
                }else {
                    result.IntersectWith(_invertedIndex.GetResultListByWord(and));
                }
            }
        }
        
        private static void AddOrWordsToResult(HashSet<string> result, List<string> orList) 
        {
            foreach (string or in orList){
                result.UnionWith(_invertedIndex.GetResultListByWord(or));
            }
        }
        
        private static void RemoveExcludeWordsFromResult(HashSet<string> result, List<string> excludeList) {
            foreach (string exclude in excludeList){
                result.ExceptWith(_invertedIndex.GetResultListByWord(exclude));
            }
        }

        private void FillListsByQuery(string query, List<string> andList, List<string> orList, List<string> excludeList)
        {
            foreach (string word in GetNormalizedString(query))
            {
                switch (word[0])
                {
                    case '+':
                        orList.Add(word.Substring(1));
                        break;
                    case '-':
                        excludeList.Add(word.Substring(1));
                        break;
                    default:
                        andList.Add(word);
                        break;
                }
            }
        }
        
        private static string[] GetNormalizedString(string query)
        {
            return query.Split(" ");
        }
        
    }
    
}

File: Phase03/src/main/java/SearchEngine.java
package main.java;

import java.util.*;

public class SearchEngine implements ISearchEngine {
    private final IInvertedIndex invertedIndex;

    public SearchEngine(IInvertedIndex invertedIndex){
        this.invertedIndex = invertedIndex;
    }

    @Override
    public Set<String> getResult(String query){

        Set<String> result = new HashSet<>();
        String modifiedQuery = getModifiedQuery(query);

        final List<String> andList = new ArrayList<>();
        final List<String> orList = new ArrayList<>();
        final List<String> excludeList = new ArrayList<>();

        fillListsByQuery(modifiedQuery, andList, orList, excludeList);

        addAndWordsToResult(result, andList);
        addOrWordsToResult(result, orList);
        removeExcludeWordsFromResult(result, excludeList);

        return result;
    }

    private String getModifiedQuery(String query) {
        return query.toLowerCase();
    }

    private void removeExcludeWordsFromResult(Set<String> result, List<String> excludeList) {
        for (String exclude : excludeList){
            result.removeAll(invertedIndex.getResultListByWord(exclude));
        }
    }

    private void addOrWordsToResult(Set<String> result, List<String> orList) {
        for (String or : orList){
            result.addAll(invertedIndex.getResultListByWord(or));
        }
    }

    private void addAndWordsToResult(Set<String> result, List<String> andList) {
        for (String and : andList){
            if (result.isEmpty()){
                result.addAll(invertedIndex.getResultListByWord(and));
            }else {
                result.retainAll(invertedIndex.getResultListByWord(and));
            }
        }
    }

    private void fillListsByQuery(String query, List<String> andList, List<String> orList, List<String> excludeList) {
        for (String word : getNormalizedString(query)){
            if(word.charAt(0)=='+'){
                orList.add(word.substring(1));
            }else if (word.charAt(0)=='-'){
                excludeList.add(word.substring(1));
            }else {
                andList.add(word);
            }
        }
    }

    private String[] getNormalizedString(String query) {
        return query.split(" ");
    }

}
